### DISCUSSION TITLE
ESP32-S3 Espressif or Synopsys DCD -- which one should I use?

### DISCUSSION THREAD

**Discussion by zjanosy** [hathach/tinyusb#1546] [Q&A] [answered]
I am a bit confused. I'm trying to make the USB Audio Class work on the ESP32-S3. I have replaced the ESP-IDF version of TinyUSB with the most recent "master" version on GitHub, because there has been quite some work on the USB Audio recently (in particular, using an explicit feedback endpoint). However, the current Espressif driver (dcd_esp32sx.c) seems to be rather incomplete, especially for isochronous transfers. I could not make it work at all.

However, I noticed that there is a Synopsys driver (dcd_dwc2.c) as well, which seems to include support for the ESP32-S3. My understanding is that the ESP32-S3 uses the same Synopsys USB controller IP as the STM32 chips, so it may work better with this driver, which seems to be more mature. I could compile and link with this driver, though it does not work either out of the box.

The question is, can I use the Synopsys driver? Why is there a separate ESP32-Sx driver, which is based on a completely different code? If I'm going to fix some bugs, which one shall I go with? As I wrote, the Synopsys driver seems to be more complete, so I would prefer it, but it would be good to know about the future of ESP32-S3 support in TinyUSB.

Thanks,
Zoltan

**Comment by hathach** [comment, marked as answer]
In short, you could actually use either dcd_esp32sx / dwc2. 

dwc2 is the newer, general and improved driver for all MCUs that use synopsys dwc2 controller (therefore it is more complicated). dcd_esp32sx is more specific to esp32sx, and does take into account of the fact that esp32sx only has 5 tx fifo but has 6 endpoints. Apart from that, it also has some tweak contributed by Espressif team. Eventually all tweaks in esp32sx will be migrated to dwc2, and it will fully replace esp32sx, for now, 90% of the case, either can be used. However for latest features such as ISO, you definitely need to use dwc2 since that is where all the future developing effort focus on.
